csi/volumeutils.py

```diff
@@ -1009,21 +1011,166 @@ def reload_glusterfs(volume):
     return True
 
 
-def mount_glusterfs(volume, mountpoint, is_client=False):
+# noqa # pylint: disable=too-many-locals
+# noqa # pylint: disable=too-many-statements
+# noqa # pylint: disable=too-few-public-methods
+# noqa # pylint: disable=too-many-branches
+class VolfileElement:
+    """Class to represent multiple elements within a volfile"""
+
+    def __init__(self, name):
+        self.name = name
+        self.type = None
+        self.options = {}
+        self.subvolumes = None
+
+
+class Volfile:
+    """Class with volfile utils methods"""
+
+    def __init__(self, volfile, elements):
+        self.volfile = volfile
+        self.elements = elements
+
+    @classmethod
+    def parse(cls, volfile):
+        """Parse volfile into multiple volfile elements"""
+
+        element = None
+        elements = []
+        with open(volfile) as volf:
+            for line in volf:
+                line = line.strip()
+                if line.startswith("volume "):
+                    element = VolfileElement(line.split()[-1])
+                elif line == "end-volume":
+                    elements.append(element)
+                    element = None
+                elif line.startswith("option "):
+                    _, name, value = line.split()
+                    element.options[name] = value
+                elif line.startswith("subvolumes "):
+                    element.subvolumes = line.split(" ", 1)[-1]
```

Comment:
- I might be wrong, does `subvolumes` always contain a single entry?
- I faintly remember there can be many clients in a subvolumes 🤔 ?

Reply:
Yes it contains multiple values. I don't think this is comes under supported volume options. Anyway @aravindavk has to confirm the list of options that will be supported by kadalu.

Reply:
it is a flat structure. Each volume entry in volfile will have only one line for subvolumes and each subvolume(Volfile subvolume) are separated by space.

Example: subvolumes replicate-spool1-0 replicate-spool1-1

Reply:
- then why are we storing only the last entry or I missed something in the parsing logic?

Reply:
Split(" ", 1) is used, all subvol list is kept unchanged and added back later.